packages/loader/container-loader/src/container.ts

```diff
@@ -1045,29 +1045,38 @@ export class Container
 	}
 
 	public dispose(error?: ICriticalContainerError): void {
-		this._deltaManager.dispose(error);
-		this.verifyClosed();
+		this.trackCloseDisposeReentry(() => this._deltaManager.dispose(error));
 	}
 
 	public close(error?: ICriticalContainerError): void {
 		// 1. Ensure that close sequence is exactly the same no matter if it's initiated by host or by DeltaManager
 		// 2. We need to ensure that we deliver disconnect event to runtime properly. See connectionStateChanged
 		//    handler. We only deliver events if container fully loaded. Transitioning from "loading" ->
 		//    "closing" will lose that info (can also solve by tracking extra state).
-		this._deltaManager.close(error);
-		this.verifyClosed();
+		this.trackCloseDisposeReentry(() => this._deltaManager.close(error));
 	}
 
-	private verifyClosed(): void {
-		assert(
-			this.connectionState === ConnectionState.Disconnected,
-			0x0cf /* "disconnect event was not raised!" */,
-		);
+	private trackCloseDisposeReentryCalls = 0;
+	private trackCloseDisposeReentry(callback: () => void): void {
+		this.trackCloseDisposeReentryCalls++;
+		try {
+			callback();
+		} finally {
+			this.trackCloseDisposeReentryCalls--;
+		}
 
-		assert(
-			this._lifecycleState === "closed" || this._lifecycleState === "disposed",
-			0x314 /* Container properly closed */,
-		);
+		// We only want to verify connectionState and lifecycleState after close/dispose has fully finished
+		if (this.trackCloseDisposeReentryCalls === 0) {
+			assert(
```

Comment:
By the way.... where do these asserts even go??  We tend to write asserts assuming they'll show up in an intuitive way, but don't always consider the callstack where the error will be thrown.

This case is especially weird because often code will intentionally close the container (rather than throwing) if they know the callstack isn't really suitable for throwing.  Or, we'll be closing because of a "top-level exception handler" of sorts (e.g. the inbound queue's "error" event closes the container).  In both these cases, it's not clear where this error will go.

Just some food for thought for our team...

In the case of the incident you faced while on call, how did the assert surface?

Reply:
Actually I'll ask this in our team chat for more visibility

Reply:
For visibility:

In the case of the assert hitting in the service load tests, this was the "close re-entry" flow:

1. Container.close() was scheduled by service load runner
2. Container.close() will "disconnect" the DeltaManager, and propagate that change to ContainerRuntime
3. ContainerRuntime.setConnectionStateCore() checks if we allow reconnection, decides we shouldn't based on number of reconnects, and calls "closeFn" (closeFn is just Container.close())
4. Re-entry Container.close() will effectively skip the DeltaManager.close() call since we're already doing it
5. 0x314 from 2nd close call
6. First close call will finish out properly and send "ContainerClose" telemetry event

Your thoughts on this make sense. Food for thought:
Where do we expect these high level asserts/errors to go? Should we add wrappers to high level API calls to ensure we catch everything we can instead of sending it up to the host to deal with?